File: packages/ts-moose-lib/src/dmv2/internal.ts
/**
 * @module internal
 * Internal implementation details for the Moose v2 data model (dmv2).
 *
 * This module manages the registration of user-defined dmv2 resources (Tables, Streams, APIs, etc.)
 * and provides functions to serialize these resources into a JSON format (`InfrastructureMap`)
 * expected by the Moose infrastructure management system. It also includes helper functions
 * to retrieve registered handler functions (for streams and APIs) and the base class
 * (`TypedBase`) used by dmv2 resource classes.
 *
 * @internal This module is intended for internal use by the Moose library and compiler plugin.
 *           Its API might change without notice.
 */
import process from "process";
import { Api, IngestApi, SqlResource, Task, Workflow } from "./index";
import { IJsonSchemaCollection } from "typia/src/schemas/json/IJsonSchemaCollection";
import { Column } from "../dataModels/dataModelTypes";
import { ClickHouseEngines, ApiUtil } from "../index";
import { OlapTable } from "./sdk/olapTable";
import { ConsumerConfig, Stream, TransformConfig } from "./sdk/stream";
import { compilerLog } from "../commons";

/**
 * Internal registry holding all defined Moose dmv2 resources.
 * Populated by the constructors of OlapTable, Stream, IngestApi, etc.
 * Accessed via `getMooseInternal()`.
 */
const moose_internal = {
  tables: new Map<string, OlapTable<any>>(),
  streams: new Map<string, Stream<any>>(),
  ingestApis: new Map<string, IngestApi<any>>(),
  apis: new Map<string, Api<any>>(),
  sqlResources: new Map<string, SqlResource>(),
  workflows: new Map<string, Workflow>(),
};
/**
 * Default retention period for streams if not specified (7 days in seconds).
 */
const defaultRetentionPeriod = 60 * 60 * 24 * 7;

/**
 * Engine-specific configuration types using discriminated union pattern
 */
interface MergeTreeEngineConfig {
  engine: "MergeTree";
}

interface ReplacingMergeTreeEngineConfig {
  engine: "ReplacingMergeTree";
}

interface AggregatingMergeTreeEngineConfig {
  engine: "AggregatingMergeTree";
}

interface SummingMergeTreeEngineConfig {
  engine: "SummingMergeTree";
}

interface S3QueueEngineConfig {
  engine: "S3Queue";
  s3Path: string;
  format: string;
  awsAccessKeyId?: string;
  awsSecretAccessKey?: string;
  compression?: string;
  headers?: { [key: string]: string };
  s3Settings?: { [key: string]: any };
}

/**
 * Union type for all supported engine configurations
 */
type EngineConfig =
  | MergeTreeEngineConfig
  | ReplacingMergeTreeEngineConfig
  | AggregatingMergeTreeEngineConfig
  | SummingMergeTreeEngineConfig
  | S3QueueEngineConfig;

/**
 * JSON representation of an OLAP table configuration.
 */
interface TableJson {
  /** The name of the table. */
  name: string;
  /** Array defining the table's columns and their types. */
  columns: Column[];
  /** List of column names used for the ORDER BY clause. */
  orderBy: string[];
  /** Engine configuration with type-safe, engine-specific parameters */
  engineConfig?: EngineConfig;
  /** Optional version string for the table configuration. */
  version?: string;
  /** Optional metadata for the table (e.g., description). */
  metadata?: { description?: string };
  /** Lifecycle management setting for the table. */
  lifeCycle?: string;
}
/**
 * Represents a target destination for data flow, typically a stream.
 */
interface Target {
  /** The name of the target resource (e.g., stream name). */
  name: string;
  /** The kind of the target resource. */
  kind: "stream"; // may add `| "table"` in the future
  /** Optional version string of the target resource's configuration. */
  version?: string;
  /** Optional metadata for the target (e.g., description for function processes). */
  metadata?: { description?: string };
}

/**
 * Represents a consumer attached to a stream.
 */
interface Consumer {
  /** Optional version string for the consumer configuration. */
  version?: string;
}

/**
 * JSON representation of a Stream/Topic configuration.
 */
interface StreamJson {
  /** The name of the stream/topic. */
  name: string;
  /** Array defining the message schema (columns/fields). */
  columns: Column[];
  /** Data retention period in seconds. */
  retentionPeriod: number;
  /** Number of partitions for the stream/topic. */
  partitionCount: number;
  /** Optional name of the OLAP table this stream automatically syncs to. */
  targetTable?: string;
  /** Optional version of the target OLAP table configuration. */
  targetTableVersion?: string;
  /** Optional version string for the stream configuration. */
  version?: string;
  /** List of target streams this stream transforms data into. */
  transformationTargets: Target[];
  /** Flag indicating if a multi-transform function (`_multipleTransformations`) is defined. */
  hasMultiTransform: boolean;
  /** List of consumers attached to this stream. */
  consumers: Consumer[];
  /** Optional description for the stream. */
  metadata?: { description?: string };
  /** Lifecycle management setting for the stream. */
  lifeCycle?: string;
}
/**
 * JSON representation of an Ingest API configuration.
 */
interface IngestApiJson {
  /** The name of the Ingest API endpoint. */
  name: string;
  /** Array defining the expected input schema (columns/fields). */
  columns: Column[];

  /** The target stream where ingested data is written. */
  writeTo: Target;
  /** The DLQ if the data does not fit the schema. */
  deadLetterQueue?: string;
  /** Optional version string for the API configuration. */
  version?: string;
  /** Optional description for the API. */
  metadata?: { description?: string };
}

/**
 * JSON representation of an API configuration.
 */
interface ApiJson {
  /** The name of the API endpoint. */
  name: string;
  /** Array defining the expected query parameters schema. */
  queryParams: Column[];
  /** JSON schema definition of the API's response body. */
  responseSchema: IJsonSchemaCollection.IV3_1;
  /** Optional version string for the API configuration. */
  version?: string;
  /** Optional description for the API. */
  metadata?: { description?: string };
}

/**
 * Represents the unique signature of an infrastructure component (Table, Topic, etc.).
 * Used for defining dependencies between SQL resources.
 */
interface InfrastructureSignatureJson {
  /** A unique identifier for the resource instance (often name + version). */
  id: string;
  /** The kind/type of the infrastructure component. */
  kind:
    | "Table"
    | "Topic"
    | "ApiEndpoint"
    | "TopicToTableSyncProcess"
    | "View"
    | "SqlResource";
}

interface WorkflowJson {
  name: string;
  retries?: number;
  timeout?: string;
  schedule?: string;
}

/**
 * JSON representation of a generic SQL resource (like View, MaterializedView).
 */
interface SqlResourceJson {
  /** The name of the SQL resource. */
  name: string;
  /** Array of SQL DDL statements required to create the resource. */
  setup: readonly string[];
  /** Array of SQL DDL statements required to drop the resource. */
  teardown: readonly string[];

  /** List of infrastructure components (by signature) that this resource reads from. */
  pullsDataFrom: InfrastructureSignatureJson[];
  /** List of infrastructure components (by signature) that this resource writes to. */
  pushesDataTo: InfrastructureSignatureJson[];
}

/**
 * Converts the internal resource registry into a structured infrastructure map.
 * This map is serialized to JSON and used by the Moose infrastructure system.
 *
 * @param registry The internal Moose resource registry (`moose_internal`).
 * @returns An object containing dictionaries of tables, topics, ingest APIs, APIs, and SQL resources, formatted according to the `*Json` interfaces.
 */
export const toInfraMap = (registry: typeof moose_internal) => {
  const tables: { [key: string]: TableJson } = {};
  const topics: { [key: string]: StreamJson } = {};
  const ingestApis: { [key: string]: IngestApiJson } = {};
  const apis: { [key: string]: ApiJson } = {};
  const sqlResources: { [key: string]: SqlResourceJson } = {};
  const workflows: { [key: string]: WorkflowJson } = {};

  registry.tables.forEach((table) => {
    // If the table is part of an IngestPipeline, inherit metadata if not set
    let metadata = (table as any).metadata;
    if (!metadata && table.config && (table as any).pipelineParent) {
      metadata = (table as any).pipelineParent.metadata;
    }
    // Create type-safe engine configuration
    const engineConfig: EngineConfig | undefined = (() => {
      switch (table.config.engine) {
        case ClickHouseEngines.MergeTree:
          return { engine: "MergeTree" };

        case ClickHouseEngines.ReplacingMergeTree:
          return { engine: "ReplacingMergeTree" };

        case ClickHouseEngines.AggregatingMergeTree:
          return { engine: "AggregatingMergeTree" };

        case ClickHouseEngines.SummingMergeTree:
          return { engine: "SummingMergeTree" };

        case ClickHouseEngines.S3Queue: {
          const s3Config = table.config as any; // Cast to access S3Queue-specific properties

          // Ensure s3Settings is initialized and has required 'mode' parameter
          const s3Settings = s3Config.s3Settings || {};
          if (!s3Settings.mode) {
            s3Settings.mode = "unordered"; // Default to 'unordered' if not specified
          }

          return {
            engine: "S3Queue",
            s3Path: s3Config.s3Path,
            format: s3Config.format,
            awsAccessKeyId: s3Config.awsAccessKeyId,
            awsSecretAccessKey: s3Config.awsSecretAccessKey,
            compression: s3Config.compression,
            headers: s3Config.headers,
            s3Settings,
          };
        }

        default:
          return undefined;
      }
    })();

    tables[table.name] = {
      name: table.name,
      columns: table.columnArray,
      orderBy: table.config.orderByFields ?? [],
      engineConfig,
      version: table.config.version,
      metadata,
      lifeCycle: table.config.lifeCycle,
    };
  });

  registry.streams.forEach((stream) => {
    // If the stream is part of an IngestPipeline, inherit metadata if not set
    let metadata = stream.metadata;
    if (!metadata && stream.config && (stream as any).pipelineParent) {
      metadata = (stream as any).pipelineParent.metadata;
    }
    const transformationTargets: Target[] = [];
    const consumers: Consumer[] = [];

    stream._transformations.forEach((transforms, destinationName) => {
      transforms.forEach(([destination, _, config]) => {
        transformationTargets.push({
          kind: "stream",
          name: destinationName,
          version: config.version,
          metadata: config.metadata,
        });
      });
    });

    stream._consumers.forEach((consumer) => {
      consumers.push({
        version: consumer.config.version,
      });
    });

    topics[stream.name] = {
      name: stream.name,
      columns: stream.columnArray,
      targetTable: stream.config.destination?.name,
      targetTableVersion: stream.config.destination?.config.version,
      retentionPeriod: stream.config.retentionPeriod ?? defaultRetentionPeriod,
      partitionCount: stream.config.parallelism ?? 1,
      version: stream.config.version,
      transformationTargets,
      hasMultiTransform: stream._multipleTransformations === undefined,
      consumers,
      metadata,
      lifeCycle: stream.config.lifeCycle,
    };
  });

  registry.ingestApis.forEach((api) => {
    // If the ingestApi is part of an IngestPipeline, inherit metadata if not set
    let metadata = api.metadata;
    if (!metadata && api.config && (api as any).pipelineParent) {
      metadata = (api as any).pipelineParent.metadata;
    }
    ingestApis[api.name] = {
      name: api.name,
      columns: api.columnArray,
      version: api.config.version,
      writeTo: {
        kind: "stream",
        name: api.config.destination.name,
      },
      deadLetterQueue: api.config.deadLetterQueue?.name,
      metadata,
    };
  });

  registry.apis.forEach((api, key) => {
    const rustKey =
      api.config.version ? `${api.name}:${api.config.version}` : api.name;
    apis[rustKey] = {
      name: api.name,
      queryParams: api.columnArray,
      responseSchema: api.responseSchema,
      version: api.config.version,
      metadata: api.metadata,
    };
  });

  registry.sqlResources.forEach((sqlResource) => {
    sqlResources[sqlResource.name] = {
      name: sqlResource.name,
      setup: sqlResource.setup,
      teardown: sqlResource.teardown,

      pullsDataFrom: sqlResource.pullsDataFrom.map((r) => {
        if (r.kind === "OlapTable") {
          const table = r as OlapTable<any>;
          const id =
            table.config.version ?
              `${table.name}_${table.config.version}`
            : table.name;
          return {
            id,
            kind: "Table",
          };
        } else if (r.kind === "SqlResource") {
          const resource = r as SqlResource;
          return {
            id: resource.name,
            kind: "SqlResource",
          };
        } else {
          throw new Error(`Unknown sql resource dependency type: ${r}`);
        }
      }),
      pushesDataTo: sqlResource.pushesDataTo.map((r) => {
        if (r.kind === "OlapTable") {
          const table = r as OlapTable<any>;
          const id =
            table.config.version ?
              `${table.name}_${table.config.version}`
            : table.name;
          return {
            id,
            kind: "Table",
          };
        } else if (r.kind === "SqlResource") {
          const resource = r as SqlResource;
          return {
            id: resource.name,
            kind: "SqlResource",
          };
        } else {
          throw new Error(`Unknown sql resource dependency type: ${r}`);
        }
      }),
    };
  });

  registry.workflows.forEach((workflow) => {
    workflows[workflow.name] = {
      name: workflow.name,
      retries: workflow.config.retries,
      timeout: workflow.config.timeout,
      schedule: workflow.config.schedule,
    };
  });

  return {
    topics,
    tables,
    ingestApis,
    apis,
    sqlResources,
    workflows,
  };
};

/**
 * Retrieves the global internal Moose resource registry.
 * Uses `globalThis` to ensure a single registry instance.
 *
 * @returns The internal Moose resource registry.
 */
export const getMooseInternal = (): typeof moose_internal =>
  (globalThis as any).moose_internal;

// work around for variable visibility in compiler output
if (getMooseInternal() === undefined) {
  (globalThis as any).moose_internal = moose_internal;
}

/**
 * Loads the user's application entry point (`app/index.ts`) to register resources,
 * then generates and prints the infrastructure map as JSON.
 *
 * This function is the main entry point used by the Moose infrastructure system
 * to discover the defined resources.
 * It prints the JSON map surrounded by specific delimiters (`___MOOSE_STUFF___start`
 * and `end___MOOSE_STUFF___`) for easy extraction by the calling process.
 */
export const dumpMooseInternal = async () => {
  loadIndex();

  console.log(
    "___MOOSE_STUFF___start",
    JSON.stringify(toInfraMap(getMooseInternal())),
    "end___MOOSE_STUFF___",
  );
};

const loadIndex = () => {
  try {
    require(`${process.cwd()}/app/index.ts`);
  } catch (error) {
    let hint: string | undefined;
    const details = error instanceof Error ? error.message : String(error);
    if (details.includes("ERR_REQUIRE_ESM") || details.includes("ES Module")) {
      hint =
        "The file or its dependencies are ESM-only. Switch to packages that dual-support CJS & ESM, or upgrade to Node 22.12+. " +
        "If you must use Node 20, you may try Node 20.19\n\n";
    }

    let options: { cause: Error } | undefined = undefined;
    if (error instanceof Error) {
      options = { cause: error };
    }

    const errorMsg = `${hint ?? ""}${details}`;
    throw new Error(errorMsg, options);
  }
};

/**
 * Loads the user's application entry point and extracts all registered stream
 * transformation and consumer functions.
 *
 * @returns A Map where keys are unique identifiers for transformations/consumers
 *          (e.g., "sourceStream_destStream_version", "sourceStream_<no-target>_version")
 *          and values are the corresponding handler functions.
 */
export const getStreamingFunctions = async () => {
  loadIndex();

  const registry = getMooseInternal();
  const transformFunctions = new Map<
    string,
    [(data: unknown) => unknown, TransformConfig<any> | ConsumerConfig<any>]
  >();

  registry.streams.forEach((stream) => {
    stream._transformations.forEach((transforms, destinationName) => {
      transforms.forEach(([_, transform, config]) => {
        const transformFunctionKey = `${stream.name}_${destinationName}${config.version ? `_${config.version}` : ""}`;
        compilerLog(`getStreamingFunctions: ${transformFunctionKey}`);
        transformFunctions.set(transformFunctionKey, [transform, config]);
      });
    });

    stream._consumers.forEach((consumer) => {
      const consumerFunctionKey = `${stream.name}_<no-target>${consumer.config.version ? `_${consumer.config.version}` : ""}`;
      transformFunctions.set(consumerFunctionKey, [
        consumer.consumer,
        consumer.config,
      ]);
    });
  });

  return transformFunctions;
};

/**
 * Loads the user's application entry point and extracts all registered
 * API handler functions.
 *
 * @returns A Map where keys are the names of the APIs and values
 *          are their corresponding handler functions.
 */
export const getApis = async () => {
  loadIndex();
  const apiFunctions = new Map<
    string,
    (params: unknown, utils: ApiUtil) => unknown
  >();

  const registry = getMooseInternal();
  // Single pass: store full keys, track aliasing decisions
  const versionCountByName = new Map<string, number>();
  const nameToSoleVersionHandler = new Map<
    string,
    (params: unknown, utils: ApiUtil) => unknown
  >();

  registry.apis.forEach((api, key) => {
    const handler = api.getHandler();
    apiFunctions.set(key, handler);

    if (!api.config.version) {
      // Explicit unversioned takes precedence for alias
      if (!apiFunctions.has(api.name)) {
        apiFunctions.set(api.name, handler);
      }
      nameToSoleVersionHandler.delete(api.name);
      versionCountByName.delete(api.name);
    } else if (!apiFunctions.has(api.name)) {
      // Only track versioned for alias if no explicit unversioned present
      const count = (versionCountByName.get(api.name) ?? 0) + 1;
      versionCountByName.set(api.name, count);
      if (count === 1) {
        nameToSoleVersionHandler.set(api.name, handler);
      } else {
        nameToSoleVersionHandler.delete(api.name);
      }
    }
  });

  // Finalize aliases for names that have exactly one versioned API and no unversioned
  nameToSoleVersionHandler.forEach((handler, name) => {
    if (!apiFunctions.has(name)) {
      apiFunctions.set(name, handler);
    }
  });

  return apiFunctions;
};

export const dlqSchema: IJsonSchemaCollection.IV3_1 = {
  version: "3.1",
  components: {
    schemas: {
      DeadLetterModel: {
        type: "object",
        properties: {
          originalRecord: {
            $ref: "#/components/schemas/Recordstringany",
          },
          errorMessage: {
            type: "string",
          },
          errorType: {
            type: "string",
          },
          failedAt: {
            type: "string",
            format: "date-time",
          },
          source: {
            oneOf: [
              {
                const: "api",
              },
              {
                const: "transform",
              },
              {
                const: "table",
              },
            ],
          },
        },
        required: [
          "originalRecord",
          "errorMessage",
          "errorType",
          "failedAt",
          "source",
        ],
      },
      Recordstringany: {
        type: "object",
        properties: {},
        required: [],
        description: "Construct a type with a set of properties K of type T",
        additionalProperties: {},
      },
    },
  },
  schemas: [
    {
      $ref: "#/components/schemas/DeadLetterModel",
    },
  ],
};

export const dlqColumns: Column[] = [
  {
    name: "originalRecord",
    data_type: "Json",
    primary_key: false,
    required: true,
    unique: false,
    default: null,
    annotations: [],
  },
  {
    name: "errorMessage",
    data_type: "String",
    primary_key: false,
    required: true,
    unique: false,
    default: null,
    annotations: [],
  },
  {
    name: "errorType",
    data_type: "String",
    primary_key: false,
    required: true,
    unique: false,
    default: null,
    annotations: [],
  },
  {
    name: "failedAt",
    data_type: "DateTime",
    primary_key: false,
    required: true,
    unique: false,
    default: null,
    annotations: [],
  },
  {
    name: "source",
    data_type: "String",
    primary_key: false,
    required: true,
    unique: false,
    default: null,
    annotations: [],
  },
];

export const getWorkflows = async () => {
  loadIndex();

  const registry = getMooseInternal();
  return registry.workflows;
};

function findTaskInTree(
  task: Task<any, any>,
  targetName: string,
): Task<any, any> | undefined {
  if (task.name === targetName) {
    return task;
  }

  if (task.config.onComplete?.length) {
    for (const childTask of task.config.onComplete) {
      const found = findTaskInTree(childTask, targetName);
      if (found) {
        return found;
      }
    }
  }

  return undefined;
}

export const getTaskForWorkflow = async (
  workflowName: string,
  taskName: string,
): Promise<Task<any, any>> => {
  const workflows = await getWorkflows();
  const workflow = workflows.get(workflowName);
  if (!workflow) {
    throw new Error(`Workflow ${workflowName} not found`);
  }

  const task = findTaskInTree(
    workflow.config.startingTask as Task<any, any>,
    taskName,
  );
  if (!task) {
    throw new Error(`Task ${taskName} not found in workflow ${workflowName}`);
  }

  return task;
};

File: packages/py-moose-lib/moose_lib/internal.py
"""
Internal utilities for Moose Python library.

This module contains Pydantic models representing the configuration signature
of various Moose resources (tables, streams/topics, APIs) and functions
to convert the user-defined resources (from `dmv2.py`) into a serializable
JSON format expected by the Moose infrastructure management system.
"""
from importlib import import_module
from typing import Literal, Optional, List, Any, Dict
from pydantic import BaseModel, ConfigDict, AliasGenerator, Field
import json
from .data_models import Column, _to_columns
from moose_lib.dmv2 import (
    get_tables,
    get_streams,
    get_ingest_apis,
    get_apis,
    get_sql_resources,
    get_workflows,
    OlapTable,
    View,
    MaterializedView,
    SqlResource
)
from pydantic.alias_generators import to_camel
from pydantic.json_schema import JsonSchemaValue

model_config = ConfigDict(alias_generator=AliasGenerator(
    serialization_alias=to_camel,
))


class Target(BaseModel):
    """Represents a target destination for data flow, typically a stream.

    Attributes:
        kind: The type of the target (currently only "stream").
        name: The name of the target stream.
        version: Optional version of the target stream configuration.
        metadata: Optional metadata for the target stream.
    """
    kind: Literal["stream"]
    name: str
    version: Optional[str] = None
    metadata: Optional[dict] = None

class Consumer(BaseModel):
    """Represents a consumer attached to a stream.

    Attributes:
        version: Optional version of the consumer configuration.
    """
    version: Optional[str] = None

class EngineConfigDict(BaseModel):
    """Engine configuration using discriminated union pattern for serialization."""
    engine: str
    # S3Queue-specific fields (only present when engine is "S3Queue")
    s3_path: Optional[str] = None
    format: Optional[str] = None
    aws_access_key_id: Optional[str] = None
    aws_secret_access_key: Optional[str] = None
    compression: Optional[str] = None
    headers: Optional[Dict[str, str]] = None
    s3_settings: Optional[Dict[str, Any]] = None

class TableConfig(BaseModel):
    """Internal representation of an OLAP table configuration for serialization.

    Attributes:
        name: Name of the table.
        columns: List of columns with their types and attributes.
        order_by: List of columns used for the ORDER BY clause.
        engine_config: Engine configuration with type-safe, engine-specific parameters.
        version: Optional version string of the table configuration.
        metadata: Optional metadata for the table.
        life_cycle: Lifecycle management setting for the table.
    """
    model_config = model_config

    name: str
    columns: List[Column]
    order_by: List[str]
    engine_config: Optional[EngineConfigDict] = None
    version: Optional[str] = None
    metadata: Optional[dict] = None
    life_cycle: Optional[str] = None

class TopicConfig(BaseModel):
    """Internal representation of a stream/topic configuration for serialization.

    Attributes:
        name: Name of the topic.
        columns: List of columns (fields) in the topic messages.
        target_table: Optional name of the OLAP table this topic automatically syncs to.
        target_table_version: Optional version of the target table configuration.
        version: Optional version string of the topic configuration.
        retention_period: Data retention period in seconds.
        partition_count: Number of partitions.
        transformation_targets: List of streams this topic transforms data into.
        has_multi_transform: Flag indicating if a multi-transform function is defined.
        consumers: List of consumers attached to this topic.
        metadata: Optional metadata for the topic.
        life_cycle: Lifecycle management setting for the topic.
    """
    model_config = model_config

    name: str
    columns: List[Column]
    target_table: Optional[str] = None
    target_table_version: Optional[str] = None
    version: Optional[str] = None
    retention_period: int
    partition_count: int
    transformation_targets: List[Target]
    has_multi_transform: bool
    consumers: List[Consumer]
    metadata: Optional[dict] = None
    life_cycle: Optional[str] = None

class IngestApiConfig(BaseModel):
    """Internal representation of an Ingest API configuration for serialization.

    Attributes:
        name: Name of the Ingest API.
        columns: List of columns expected in the input data.
        write_to: The target stream where the ingested data is written.
        version: Optional version string of the API configuration.
        metadata: Optional metadata for the API.
    """
    model_config = model_config

    name: str
    columns: List[Column]
    write_to: Target
    dead_letter_queue: Optional[str] = None
    version: Optional[str] = None
    metadata: Optional[dict] = None

class InternalApiConfig(BaseModel):
    """Internal representation of a API configuration for serialization.

    Attributes:
        name: Name of the API.
        query_params: List of columns representing the expected query parameters.
        response_schema: JSON schema definition of the API's response body.
        version: Optional version string of the API configuration.
        metadata: Optional metadata for the API.
    """
    model_config = model_config

    name: str
    query_params: List[Column]
    response_schema: JsonSchemaValue
    version: Optional[str] = None
    metadata: Optional[dict] = None

class WorkflowJson(BaseModel):
    """Internal representation of a workflow configuration for serialization.

    Attributes:
        name: Name of the workflow.
        retries: Optional number of retry attempts for the entire workflow.
        timeout: Optional timeout string for the entire workflow.
        schedule: Optional cron-like schedule string for recurring execution.
    """
    model_config = model_config

    name: str
    retries: Optional[int] = None
    timeout: Optional[str] = None
    schedule: Optional[str] = None

class InfrastructureSignatureJson(BaseModel):
    """Represents the unique signature of an infrastructure component (Table, Topic, etc.).

    Used primarily for defining dependencies between SQL resources.

    Attributes:
        id: A unique identifier for the resource instance (often name + version).
        kind: The type of the infrastructure component.
    """
    id: str
    kind: Literal["Table", "Topic", "ApiEndpoint", "TopicToTableSyncProcess", "View", "SqlResource"]

class SqlResourceConfig(BaseModel):
    """Internal representation of a generic SQL resource (like View, MaterializedView) for serialization.

    Attributes:
        name: Name of the SQL resource.
        setup: List of SQL commands required to create the resource.
        teardown: List of SQL commands required to drop the resource.
        pulls_data_from: List of infrastructure components this resource reads from.
        pushes_data_to: List of infrastructure components this resource writes to.
        metadata: Optional metadata for the resource.
    """
    model_config = model_config

    name: str
    setup: list[str]
    teardown: list[str]
    pulls_data_from: list[InfrastructureSignatureJson]
    pushes_data_to: list[InfrastructureSignatureJson]
    metadata: Optional[dict] = None


class InfrastructureMap(BaseModel):
    """Top-level model holding the configuration for all defined Moose resources.

    This structure is serialized to JSON and passed to the Moose infrastructure system.

    Attributes:
        tables: Dictionary mapping table names to their configurations.
        topics: Dictionary mapping topic/stream names to their configurations.
        ingest_apis: Dictionary mapping ingest API names to their configurations.
        apis: Dictionary mapping API names to their configurations.
        sql_resources: Dictionary mapping SQL resource names to their configurations.
        workflows: Dictionary mapping workflow names to their configurations.
    """
    model_config = model_config

    tables: dict[str, TableConfig]
    topics: dict[str, TopicConfig]
    ingest_apis: dict[str, IngestApiConfig]
    apis: dict[str, InternalApiConfig]
    sql_resources: dict[str, SqlResourceConfig]
    workflows: dict[str, WorkflowJson]


def _map_sql_resource_ref(r: Any) -> InfrastructureSignatureJson:
    """Maps a `dmv2` SQL resource object to its `InfrastructureSignatureJson`.

    Determines the correct `kind` and generates the `id` based on the resource
    type and its configuration (e.g., including version if present).

    Args:
        r: An instance of OlapTable, View, MaterializedView, or SqlResource.

    Returns:
        An InfrastructureSignatureJson representing the resource.

    Raises:
        TypeError: If the input object is not a recognized SQL resource type.
    """
    if hasattr(r, 'kind'):
        if r.kind == "OlapTable":
            # Explicitly cast for type hint checking if needed, though Python is dynamic
            table = r # type: OlapTable
            res_id = f"{table.name}_{table.config.version}" if table.config.version else table.name
            return InfrastructureSignatureJson(id=res_id, kind="Table")
        elif r.kind == "SqlResource":
            # Explicitly cast for type hint checking if needed
            resource = r # type: SqlResource
            return InfrastructureSignatureJson(id=resource.name, kind="SqlResource")
        else:
            raise TypeError(f"Unknown SQL resource kind: {r.kind} for object: {r}")
    else:
        # Fallback or error if 'kind' attribute is missing
        raise TypeError(f"Object {r} lacks a 'kind' attribute for dependency mapping.")


def _convert_engine_to_config_dict(engine_enum, table) -> EngineConfigDict:
    """Convert engine enum and table configuration to new engine config format.
    
    Args:
        engine_enum: The ClickHouseEngines enum value
        table: The OlapTable instance with configuration
        
    Returns:
        EngineConfigDict with engine-specific configuration
    """
    from moose_lib import ClickHouseEngines
    from moose_lib.blocks import S3QueueEngine
    
    # Check if the table uses the new engine configuration classes
    if hasattr(table.config, 'engine') and hasattr(table.config.engine, '__class__'):
        engine_instance = table.config.engine
        if isinstance(engine_instance, S3QueueEngine):
            return EngineConfigDict(
                engine="S3Queue",
                s3_path=engine_instance.s3_path,
                format=engine_instance.format,
                aws_access_key_id=engine_instance.aws_access_key_id,
                aws_secret_access_key=engine_instance.aws_secret_access_key,
                compression=engine_instance.compression,
                headers=engine_instance.headers,
                s3_settings=engine_instance.s3_settings
            )
    
    # Handle legacy enum-based engine configuration
    engine_name = engine_enum.value if hasattr(engine_enum, 'value') else str(engine_enum)
    
    # For S3Queue with legacy configuration, check for s3_queue_engine_config
    if engine_name == "S3Queue" and hasattr(table.config, 's3_queue_engine_config'):
        s3_config = table.config.s3_queue_engine_config
        if s3_config:
            return EngineConfigDict(
                engine="S3Queue",
                s3_path=s3_config.path,
                format=s3_config.format,
                aws_access_key_id=s3_config.aws_access_key_id,
                aws_secret_access_key=s3_config.aws_secret_access_key,
                compression=s3_config.compression,
                headers=s3_config.headers,
                s3_settings=s3_config.settings
            )
    
    # For all other engines, just return the engine name
    return EngineConfigDict(
        engine=engine_name,
        s3_path=None,
        aws_access_key_id=None,
        aws_secret_access_key=None,
        s3_settings=None
    )

def to_infra_map() -> dict:
    """Converts the registered `dmv2` resources into the serializable `InfrastructureMap` format.

    Iterates through the internal registries (`_tables`, `_streams`, etc.) populated
    by the user's definitions in `app/main.py` (or elsewhere) and transforms them
    into the corresponding `*Config` Pydantic models.

    Returns:
        A dictionary representing the `InfrastructureMap`, ready for JSON serialization
        using Pydantic's `model_dump` with camelCase aliases.
    """
    tables = {}
    topics = {}
    ingest_apis = {}
    apis = {}
    sql_resources = {}
    workflows = {}

    for name, table in get_tables().items():
        # Convert engine configuration to new format
        engine_config = None
        if table.config.engine:
            engine_config = _convert_engine_to_config_dict(table.config.engine, table)
        
        tables[name] = TableConfig(
            name=name,
            columns=_to_columns(table._t),
            order_by=table.config.order_by_fields,
            engine_config=engine_config,
            version=table.config.version,
            metadata=getattr(table, "metadata", None),
            life_cycle=table.config.life_cycle.value if table.config.life_cycle else None,
        )

    for name, stream in get_streams().items():
        transformation_targets = [
            Target(
                kind="stream",
                name=dest_name,
                version=transform.config.version,
                metadata=getattr(transform.config, "metadata", None),
            )
            for dest_name, transforms in stream.transformations.items()
            for transform in transforms
        ]

        consumers = [
            Consumer(version=consumer.config.version)
            for consumer in stream.consumers
        ]

        topics[name] = TopicConfig(
            name=name,
            columns=_to_columns(stream._t),
            target_table=stream.config.destination.name if stream.config.destination else None,
            target_table_version=stream.config.destination.config.version if stream.config.destination else None,
            retention_period=stream.config.retention_period,
            partition_count=stream.config.parallelism,
            version=stream.config.version,
            transformation_targets=transformation_targets,
            has_multi_transform=stream._multipleTransformations is not None,
            consumers=consumers,
            metadata=getattr(stream, "metadata", None),
            life_cycle=stream.config.life_cycle.value if stream.config.life_cycle else None,
        )

    for name, api in get_ingest_apis().items():
        ingest_apis[name] = IngestApiConfig(
            name=name,
            columns=_to_columns(api._t),
            version=api.config.version,
            write_to=Target(
                kind="stream",
                name=api.config.destination.name
            ),
            metadata=getattr(api, "metadata", None),
            dead_letter_queue=api.config.dead_letter_queue.name if api.config.dead_letter_queue else None
        )

    for name, api in get_apis().items():
        apis[name] = InternalApiConfig(
            name=api.name,
            query_params=_to_columns(api.model_type),
            response_schema=api.get_response_schema(),
            version=api.config.version,
            metadata=getattr(api, "metadata", None),
        )

    for name, resource in get_sql_resources().items():
        sql_resources[name] = SqlResourceConfig(
            name=resource.name,
            setup=resource.setup,
            teardown=resource.teardown,
            pulls_data_from=[_map_sql_resource_ref(dep) for dep in resource.pulls_data_from],
            pushes_data_to=[_map_sql_resource_ref(dep) for dep in resource.pushes_data_to],
            metadata=getattr(resource, "metadata", None),
        )

    for name, workflow in get_workflows().items():
        workflows[name] = WorkflowJson(
            name=workflow.name,
            retries=workflow.config.retries,
            timeout=workflow.config.timeout,
            schedule=workflow.config.schedule,
        )

    infra_map = InfrastructureMap(
        tables=tables,
        topics=topics,
        ingest_apis=ingest_apis,
        apis=apis,
        sql_resources=sql_resources,
        workflows=workflows
    )

    return infra_map.model_dump(by_alias=True)


def load_models():
    """Imports the user's main application module and prints the infrastructure map.

    This function is typically the entry point for the Moose infrastructure system
    when processing Python-defined resources.

    1. Imports `app.main`, which should trigger the registration of all Moose
       resources defined therein (OlapTable[...](...), Stream[...](...), etc.).
    2. Calls `to_infra_map()` to generate the infrastructure configuration dictionary.
    3. Prints the dictionary as a JSON string, wrapped in specific delimiters
       (`___MOOSE_STUFF___start` and `end___MOOSE_STUFF___`), which the
       calling system uses to extract the configuration.
    """
    import_module("app.main")

    # Generate the infrastructure map
    infra_map = to_infra_map()

    # Print in the format expected by the infrastructure system
    print("___MOOSE_STUFF___start", json.dumps(infra_map), "end___MOOSE_STUFF___")
